traits/trait_list_object.py

```diff
@@ -53,460 +88,750 @@ def __init__(self, index=0, removed=None, added=None):
             return
 
-        removed = self.copy()
-        list.clear(self)
+        object = self.object()
+        if object is None:
+            return
 
-        if self.name_items is not None:
-            self._send_trait_items_event(
-                self.name_items, TraitListEvent(0, removed)
-            )
+        # bug-for-bug conversion of parameters to TraitListEvent
+        if isinstance(index, slice):
+            if index.step in {1, None}:
+                index = min(index.start, index.stop)
+            else:
+                if added and not isinstance(added, list):
+                    added = [added]
+                if not isinstance(removed, list):
+                    removed = [removed]
+        else:
+            if removed is Undefined:
+                removed = []
+            elif not isinstance(removed, list):
+                removed = [removed]
+            if added is Undefined:
+                added = []
+            elif not isinstance(added, list):
+                added = [added]
+        event = TraitListEvent(index, removed, added)
+        items_event = self.trait.items_event()
+        object.trait_items_event(self.name_items, event, items_event)
 
-    def len_error(self, len):
-        """ Utility method that raises an error if length is incorrect. """
-        raise TraitError(
-            "The '%s' trait of %s instance must be %s, "
-            "but you attempted to change its length to %d element%s."
-            % (
-                self.name,
-                class_of(self.object()),
-                self.trait.full_info(self.object(), self.name, Undefined),
-                len,
-                "s"[len == 1:],
-            )
+    def __deepcopy__(self, memo):
+        """ Perform a deepcopy operation.
+
+        Notifiers are transient and should not be copied.
+        """
+        id_self = id(self)
+        if id_self in memo:
+            return memo[id_self]
+
+        memo[id_self] = result = TraitListObject(
+            self.trait,
+            lambda: None,
+            self.name,
+            [copy.deepcopy(x, memo) for x in self],
         )
 
+        return result
+
     def __getstate__(self):
+        """ Get the state of the object for serialization.
+
+        Notifiers are transient and should not be serialized.
+        """
         result = self.__dict__.copy()
         result.pop("object", None)
         result.pop("trait", None)
 
         return result
 
     def __setstate__(self, state):
+        """ Restore the state of the object after serialization.
+
+        Notifiers are transient and are restored to the empty list.
+        """
         name = state.setdefault("name", "")
         object = state.pop("object", None)
         if object is not None:
-            self.object = ref(object)
-            self.rename(name)
+            state['object'] = ref(object)
+            trait = self.object()._trait(name, 0)
```

Comment:
Before setting the state, what is `self.object` and is it the value we should use?  Do we have a test for this?

Reply:
`self.object` will be `ref(object)` which has the trait. I'm not sure about the test,  would just pickle and unpickle a `List` be enough ?

Reply:
Yes, with a value for `object` that isn't None.
The following test currently fails with an AttributeError that can be fixed by changing `self.object` here to `object`:
```

class Foo(HasTraits):

    my_list = List()

...
    def test_pickle_with_object(self):
        foo = Foo(my_list=[1, 2, 3])
        serialized = pickle.dumps(foo)

        foo_unpickled = pickle.loads(serialized)

        self.assertIs(foo_unpickled.my_list.object(), foo_unpickled)
        self.assertIsInstance(foo_unpickled.my_list.trait, List)
```

Reply:
Am I missing something? This test is passing for me without any changes to the current code .

Reply:
Hmm that is strange, now I also can't reproduce the test failure. However the code branch seems to not be covered by this test nor the existing tests.  I am still not convinced this is correct!

Reply:
I am now convinced that this block of code is dead.
It was from 12 years ago: https://github.com/enthought/traits/commit/d1e3bb8c275da271aee621f3906095228fe53495
It is made dead for traits 4.0.0 because `object` is removed in `__getstate__`:  https://github.com/enthought/traits/commit/4163f83aea1c05308687102e6a7a280d16924d95

While this block of code looks wrong, we don't need to fix it, it is going to go.
Resolving as won't fix.

(still puzzled by my test failure earlier, no idea what was going on there.)